File: src/brs/assetexchange/AssetExchange.java
package brs.assetexchange;

import brs.Account.AccountAsset;
import brs.*;
import brs.Attachment.ColoredCoinsAskOrderPlacement;
import brs.Attachment.ColoredCoinsAssetIssuance;
import brs.Attachment.ColoredCoinsBidOrderPlacement;
import brs.Order.Ask;
import brs.Order.Bid;
import brs.Order.OrderJournal;
import brs.Trade.Event;
import brs.util.CollectionWithIndex;
import brs.util.Listener;

public interface AssetExchange {

  CollectionWithIndex<Asset> getAllAssets(int from, int to);

  CollectionWithIndex<Asset> getAssetsByName(String name, int from, int to);

  Asset getAsset(long assetId);

  long getTradeVolume(long assetId, int heightStart, int heightEnd);

  long getHighPrice(long assetId, int heightStart, int heightEnd);

  long getLowPrice(long assetId, int heightStart, int heightEnd);

  long getOpenPrice(long assetId, int heightStart, int heightEnd);

  long getClosePrice(long assetId, int heightStart, int heightEnd);

  int getTradeCount(long assetId);

  int getTransferCount(long id);

  int getAssetAccountsCount(Asset asset, long minimumQuantity, boolean ignoreTreasury, boolean unconfirmed);

  long getAssetCirculatingSupply(Asset asset, boolean ignoreTreasury, boolean unconfirmed);

  void addTradeListener(Listener<Trade> listener, Event trade);

  Ask getAskOrder(long orderId);

  void addAsset(long assetId, long accountId, ColoredCoinsAssetIssuance attachment);

  void addAssetTransfer(Transaction transaction, long assetId, long quantityQNT);

  void addAskOrder(Transaction transaction, ColoredCoinsAskOrderPlacement attachment);

  void addBidOrder(Transaction transaction, ColoredCoinsBidOrderPlacement attachment);

  void removeAskOrder(long orderId);

  Order.Bid getBidOrder(long orderId);

  void removeBidOrder(long orderId);

  CollectionWithIndex<Trade> getAllTrades(int from, int to);

  CollectionWithIndex<Trade> getTrades(long assetId, int from, int to);

  CollectionWithIndex<Trade> getAccountTrades(long accountId, int from, int to);

  CollectionWithIndex<Trade> getAccountAssetTrades(long accountId, long assetId, int from, int to);
  
  CollectionWithIndex<OrderJournal> getOrderJournal(long assetId, long accountId, int from, int to);

  CollectionWithIndex<AccountAsset> getAssetAccounts(Asset asset, boolean ignoreTreasury, long minimumQuantity, boolean unconfirmed, int from, int to);

  CollectionWithIndex<Asset> getAssetsIssuedBy(long accountId, int from, int to);

  CollectionWithIndex<Asset> getAssetsOwnedBy(long accountId, int from, int to);

  CollectionWithIndex<AssetTransfer> getAssetTransfers(long assetId, int from, int to);

  CollectionWithIndex<AssetTransfer> getAccountAssetTransfers(long id, long id1, int from, int to);

  int getAssetsCount();

  int getAskCount();

  int getBidCount();

  int getTradesCount();

  int getAssetTransferCount();

  CollectionWithIndex<Ask> getAskOrdersByAccount(long accountId, int from, int to);

  CollectionWithIndex<Ask> getAskOrdersByAccountAsset(long accountId, long assetId, int from, int to);

  CollectionWithIndex<Bid> getBidOrdersByAccount(long accountId, int from, int to);

  CollectionWithIndex<Bid> getBidOrdersByAccountAsset(long accountId, long assetId, int from, int to);

  CollectionWithIndex<Ask> getAllAskOrders(int from, int to);

  CollectionWithIndex<Bid> getAllBidOrders(int from, int to);

  CollectionWithIndex<Ask> getSortedAskOrders(long assetId, int from, int to);

  CollectionWithIndex<Bid> getSortedBidOrders(long assetId, int from, int to);

}

File: src/main/java/brs/assetexchange/AssetExchange.kt
package brs.assetexchange

import brs.Account.AccountAsset
import brs.Asset
import brs.AssetTransfer
import brs.Attachment.*
import brs.Order.Ask
import brs.Order.Bid
import brs.Trade
import brs.Trade.Event
import brs.Transaction

interface AssetExchange {
    suspend fun getAssetsCount(): Int

    suspend fun getAskCount(): Int

    suspend fun getBidCount(): Int

    suspend fun getTradesCount(): Int

    suspend fun getAssetTransferCount(): Int

    suspend fun getAllAssets(from: Int, to: Int): Collection<Asset>

    suspend fun getAsset(assetId: Long): Asset?

    suspend fun getTradeCount(assetId: Long): Int

    suspend fun getTransferCount(assetId: Long): Int

    suspend fun getAssetAccountsCount(assetId: Long): Int

    suspend fun addTradeListener(eventType: Event, listener: suspend (Trade) -> Unit)

    suspend fun getAskOrder(orderId: Long): Ask?

    suspend fun addAsset(transaction: Transaction, attachment: ColoredCoinsAssetIssuance)

    suspend fun addAssetTransfer(transaction: Transaction, attachment: ColoredCoinsAssetTransfer)

    suspend fun addAskOrder(transaction: Transaction, attachment: ColoredCoinsAskOrderPlacement)

    suspend fun addBidOrder(transaction: Transaction, attachment: ColoredCoinsBidOrderPlacement)

    suspend fun removeAskOrder(orderId: Long)

    suspend fun getBidOrder(orderId: Long): Bid?

    suspend fun removeBidOrder(orderId: Long)

    suspend fun getAllTrades(i: Int, i1: Int): Collection<Trade>

    suspend fun getTrades(assetId: Long, from: Int, to: Int): Collection<Trade>

    suspend fun getAccountTrades(accountId: Long, from: Int, to: Int): Collection<Trade>

    suspend fun getAccountAssetTrades(accountId: Long, assetId: Long, from: Int, to: Int): Collection<Trade>

    suspend fun getAccountAssetsOverview(accountId: Long, height: Int, from: Int, to: Int): Collection<AccountAsset>

    suspend fun getAssetsIssuedBy(accountId: Long, from: Int, to: Int): Collection<Asset>

    suspend fun getAssetTransfers(assetId: Long, from: Int, to: Int): Collection<AssetTransfer>

    suspend fun getAccountAssetTransfers(id: Long, id1: Long, from: Int, to: Int): Collection<AssetTransfer>

    suspend fun getAskOrdersByAccount(accountId: Long, from: Int, to: Int): Collection<Ask>

    suspend fun getAskOrdersByAccountAsset(accountId: Long, assetId: Long, from: Int, to: Int): Collection<Ask>

    suspend fun getBidOrdersByAccount(accountId: Long, from: Int, to: Int): Collection<Bid>

    suspend fun getBidOrdersByAccountAsset(accountId: Long, assetId: Long, from: Int, to: Int): Collection<Bid>

    suspend fun getAllAskOrders(from: Int, to: Int): Collection<Ask>

    suspend fun getAllBidOrders(from: Int, to: Int): Collection<Bid>

    suspend fun getSortedAskOrders(assetId: Long, from: Int, to: Int): Collection<Ask>

    suspend fun getSortedBidOrders(assetId: Long, from: Int, to: Int): Collection<Bid>
}
